### PR TITLE
Task/subscription mechanism

### PR DESCRIPTION
**Pull request by dmoranj**
Implements #259

### CODE REVIEW

File: test/unit/subscriptionRequests/simpleNotification.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        "isPattern": "false",
+        "id": "FirstMicroLight"
+      },
+      "statusCode": {
+        "code": "200",
+        "reasonPhrase": "OK"
+      }
+    }
+  ]
+}
```
Comment (AlvaroVega) [inline test/unit/subscriptionRequests/simpleNotification.json]: no new line at the end of the file

File: lib/templates/notificationTemplate.json [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    "subscriptionId": {
+      "type": "string",
+      "required": true
+    },
+    "originator": {
+      "type": "string",
+      "required": true
+    }
+  }
+}
```
Comment (gtorodelvalle) [inline lib/templates/notificationTemplate.json]: Final blank line appreciated... :)

File: test/unit/subscriptionRequests/errorNotification.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        "isPattern": "false",
+        "id": "FirstMicroLight"
+      },
+      "statusCode": {
+        "code": "500",
+        "reasonPhrase": "NOK"
+      }
+    }
+  ]
+}
```
Comment (gtorodelvalle) [inline test/unit/subscriptionRequests/errorNotification.json]: Final blank line appreciated... :)

File: test/unit/subscriptionResponses/simpleSubscriptionSuccess.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "subscribeResponse": {
+    "duration": "P1M",
+    "subscriptionId": "51c0ac9ed714fb3b37d7d5a8"
+  }
+}
```
Comment (gtorodelvalle) [inline test/unit/subscriptionResponses/simpleSubscriptionSuccess.json]: Final blank line appreciated... :)

Comment (gtorodelvalle) [inline test/unit/subscriptionResponses/simpleSubscriptionSuccess.json]: LGTM on my side! ;)

File: lib/services/ngsi/subscriptionService.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        headers: {
+            'fiware-service': device.service,
+            'fiware-servicepath': device.subservice
+        },
+        json: {
+            subscriptionId: id,
+            duration: 'P0S'
+        }
+
+    };
```
Comment (anabelengp) [inline lib/services/ngsi/subscriptionService.js]: Wouldn't it be better to use either POST /v1/unsubscribeContext or convop DELETE /v1/contextSubscriptions/{subscriptionID}? I think both of them remove the subscription form orion DB. (At least the convop does)

File: lib/services/ngsi/subscriptionService.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            ],
+            reference: config.providerUrl + '/notify',
+            duration: config.deviceRegistrationDuration || 'P1M',
+            notifyConditions: [
+                {
+                    type: 'ONCHANGE',
+                    condValues: triggers
+                }
+            ],
+            throttling: config.throttling || 'PT5S'
```
Comment (anabelengp) [inline lib/services/ngsi/subscriptionService.js]: Does the IOTA need throttilng to be controlled? It config.throttling is not set, PT5S could be harmful...
From my experience this two cases were problematic if I used throttling:
* UpdateContext to APPEND an entity and then another  request to UPDATE the same entity. (Perhaps, Throttling here could be desirable)
* UpdateContext to APPEND **several** entities. This produces two consecutive notifications and if throttling is set the second one can be ignored.

File: lib/services/ngsi/subscriptionService.js (hunk 5 above)
Comment (dmoranj) [inline lib/services/ngsi/subscriptionService.js]: I didn't know that functionality was already implemented in Orion. I'll change it.

File: lib/services/ngsi/subscriptionService.js (hunk 6 above)
Comment (dmoranj) [inline lib/services/ngsi/subscriptionService.js]: I don't have a strong opinion on that matter. Let's discuss it offline and I will remove the default value if it seems problematic.

File: test/unit/subscriptionRequests/simpleNotification.json (hunk 1 above)
Comment (dmoranj) [inline test/unit/subscriptionRequests/simpleNotification.json]: Fixed in ad8cfe4

File: lib/templates/notificationTemplate.json (hunk 2 above)
Comment (dmoranj) [inline lib/templates/notificationTemplate.json]: Fixed in ad8cfe4

File: test/unit/subscriptionRequests/errorNotification.json (hunk 3 above)
Comment (dmoranj) [inline test/unit/subscriptionRequests/errorNotification.json]: Fixed in ad8cfe4

File: test/unit/subscriptionResponses/simpleSubscriptionSuccess.json (hunk 4 above)
Comment (dmoranj) [inline test/unit/subscriptionResponses/simpleSubscriptionSuccess.json]: Fixed in ad8cfe4

**Comment by XavierVal** [comment]
LGTM

**Comment by anabelengp** [comment]
LGTM

**Comment by mrutid** [comment]
ok

**Pull request merged**